keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+
+One alternative would be to have the CSI driver publish its capabilities through the `CSIDriver` object. This could be done in two ways
+* The CSI driver author would manually put the list of capabilities in the [`CSIPlugin` object](https://github.com/kubernetes-csi/csi-driver-host-path/blob/3d7b5eac9659bb2d436d4ed4d17b10020bc2677c/deploy/kubernetes-1.24/hostpath/csi-hostpath-driverinfo.yaml).
+* We'd have a plugin in the [driver's deployment](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/deploy/kubernetes-1.24/hostpath/csi-hostpath-plugin.yaml) that would automatically populate the `CSIDriver` object with the capabilities of the driver by calling `GetPluginCapabilities`.
+  * We could also modify one of the existing plugins (eg `csi-resizer`, `csi-provisioner`) - but none of them looks like a good candidate for this purpose.
+
+The two options have the same drawbacks
+
+* It will require modifying existing CSI drivers to support this feature.
+* I noticed that some drivers don't explicitly specify support for online expansion when calling `GetPluginCapabilities`. Here's an example from the [EBS CSI Driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/7db7d1a745abd1473eb86e63383c119209dcfc80/pkg/driver/identity.go#L36)
+
+Given the above, this kind of validation won't be handled as part of this KEP.
+
+#### Validating That The Volume Size Doesn't Decrease
+
+We don't want to prevent users from decreasing the size of the volume in the StatefulSet PVC templates. Because, we want to support recovery from expansion failures as discussed in [KEP 1790](../1790-recover-resize-failure/README.md). 
+
+However, we'd ideally want to prevent users from decreasing the size of the StatefulSet PVC templates below (or equal) to the current capacity of the volume (`.status.capacity.storage`). 
+This is similar to a validation that performed by the [PVC API server validation code](https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/apis/core/validation/validation.go#L2256) today. 
+
+There are two problems with performing this kind of validation during StatefulSet admission though.
+
+* The validation logic will have to be informed of and access all of the different PVC objects that exist for the StatefulSet. This might involve a lot of logic that is owned by the StatefulSet controller today (eg PVC naming convensions, ordinals, etc)
+* There could be a race condition where at the time of admission, we see the requested size as being less than the capacity of a PVC - but by the time that the StatefulSet controller tries to reconcole the PVC's request size to that of the template, the capabity of the PVC has already increased.
+
+Given the above, this kind of validation won't be handled as part of this KEP.
```

Comment:
If we can't validate if it is safe to reduce size to certain value, then statefulset controller will keep changing PVC size to a value that will be most likely be rejected by API server.

Reply:
I should probably remove this point

> The validation logic will have to be informed of and access all of the different PVC objects that exist for the StatefulSet. This might involve a lot of logic that is owned by the StatefulSet controller today (eg PVC naming convensions, ordinals, etc)

because by adding an high watermark value of all the PVCs capacities to the statefulset status field (that will be updated during reconciliation), it should be easy, I think, to have a validation logic that ensures that the user doesn't resize to a size lower than the PVC with the highest capacity.
But again, this is still vulnurable to race conditions (eg the watermark changes between validation and next statefulset reconciliation)

But we can still have a best effort validation, knowing that there is a posibility of a race condition. if the race condition does happen, the user would have the option to resize the PVC template back up

wdyt?